Python/sysmodule.c

```diff
@@ -314,28 +313,79 @@ sys_audit_tstate(PyThreadState *ts, const char *event,
     return res;
 }
 
+static int
+sys_audit_vargs(PyThreadState *tstate, const char *event,
+                const char *format, va_list vargs)
+{
+    if (format && strchr(format, 'N')) {
+        _PyErr_SetString(tstate, PyExc_ValueError,
+                         "format argument must not use the 'N' format");
+        return -1;
+    }
+
+    PyObject *args;
+    if (format && format[0]) {
+        args = Py_VaBuildValue(format, vargs);
+        if (args && !PyTuple_Check(args)) {
+            PyObject *argTuple = PyTuple_Pack(1, args);
+            Py_SETREF(args, argTuple);
+        }
+    }
+    else {
+        args = PyTuple_New(0);
+    }
+    if (!args) {
+        return -1;
+    }
+
+    int res = sys_audit_tuple(tstate, event, args);
+    Py_DECREF(args);
+    return res;
+}
+
 int
 _PySys_Audit(PyThreadState *tstate, const char *event,
-             const char *argFormat, ...)
+             const char *format, ...)
 {
+    // tstate can be NULL
+    if (!should_audit_tstate(tstate)) {
+        return 0;
+    }
+
     va_list vargs;
-    va_start(vargs, argFormat);
-    int res = sys_audit_tstate(tstate, event, argFormat, vargs);
+    va_start(vargs, format);
+    int res = sys_audit_vargs(tstate, event, format, vargs);
     va_end(vargs);
     return res;
 }
 
 int
-PySys_Audit(const char *event, const char *argFormat, ...)
+PySys_Audit(const char *event, const char *format, ...)
 {
     PyThreadState *tstate = _PyThreadState_GET();
+    // tstate can be NULL if the function is called before Py_Initialize()
+    if (!should_audit_tstate(tstate)) {
+        return 0;
+    }
+
     va_list vargs;
-    va_start(vargs, argFormat);
-    int res = sys_audit_tstate(tstate, event, argFormat, vargs);
+    va_start(vargs, format);
+    int res = sys_audit_vargs(tstate, event, format, vargs);
     va_end(vargs);
     return res;
 }
 
+int
+PySys_AuditTuple(const char *event, PyObject *args)
+{
+    if (args) {
```

Comment:
You need to check that `args` is a tuple, otherwise `PySys_AuditTuple()` will return successfully, but doing not what was expected.

Reply:
> You need to check that args is a tuple

I implemented the change requested by @zooba: https://github.com/python/cpython/pull/108965#issuecomment-1709022843

> otherwise PySys_AuditTuple() will return successfully, but doing not what was expected.

I understood that it's the wanted behavior according to @zooba: https://github.com/python/cpython/pull/108965#issuecomment-1709232315

Did I misunderstood @zooba?

Reply:
I thought that you misunderstood @zooba, but I can misunderstood him. I understood it that the magic is in "unpacking a tuple" if the format is a single "O". The magic can be cancelled by using format "(O)".

But maybe I misunderstood the design of `PySys_AuditTuple()`. Is it intentional to auto-pack non-tuple argument into a 1-tuple? It may be convenient if you have many audit events with a single argument, but how much is it common?

Reply:
Hum ok, I changed my PR to require args to be a tuple, as I did in my first version.